accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,190 @@
+
+The API surface clearly delineates buffers for which the caller maintains responsibility (the `ctor` and `CopyTo` methods). If the caller wishes that the plaintext data is never copied during a GC, then the destination buffer they pass to `CopyTo` should be allocated from the pinned object heap, the stack, or the native memory pool. The API surface gives them full control over that process.
+
+```cs
+// Example showing unshrouding a ShroudedBuffer<char> into a temp char[].
+ShroudedBuffer<char> shroudedBuffer = GetShroudedBuffer();
+char[] tempArray = GC.AllocateArray<char>(shroudedBuffer.Length, pinned: true);
+try
+{
+    shroudedBuffer.CopyTo(tempArray);
+    // use 'tempArray' here
+}
+finally
+{
+    Array.Clear(tempArray);
+}
+
+// Example showing unshrouding a ShroudedBuffer<char> into a string.
+ShroudedBuffer<char> shroudedBuffer = GetShroudedBuffer();
+string unshroudedString = string.Create(
+    shroudedBuffer.Length, shroudedBuffer, (span, shroudedBuffer) =>
+    {
+        shroudedBuffer.CopyTo(span);
+    });
+```
+
+We would clearly document that this type is not intended to offer cryptographic protection of the buffer contents; it's simply intended to enforce more hygienic usage of data by callers.
+
+The type also intentionally does not give callers access to the raw backing span. This is because how the backing data is stored is strictly an implementation detail (it might not even be in-proc!), and exposing the span would leak that detail into the public API surface. The `CopyTo` method is the only way to get the raw contents. This also means that such buffers are immutable once created.
+
+__We should take care not to position this "shrouded buffer" type as a security feature.__ It's intended solely for code hygiene purposes.
+
+This proposal intentionally excludes `ToString`, `ToArray`, or other methods which might unshroud the buffer. `ToString` is called frequently by library code as part of logging, and we always want unshrouding to be a _deliberate_ action. Additionally, we want the caller to maintain full control over the unshrouded buffer where the contents are written, which precludes "opinionated" APIs like `ToArray` which create such buffers themselves.
+
+The proposal also intentionally leaves the type unsealed but provides no virtual methods other than `Dispose`. This is because all aspects of the implementation must be free to change between .NET versions. If there is a desire to modify the implemenation itself, we could consider an interface `IShroudedBuffer<T>` to accompany the concrete type.
```

Comment:
```suggestion
The proposal also intentionally leaves the type unsealed but provides no virtual methods other than `Dispose`. This is because all aspects of the implementation must be free to change between .NET versions. If there is a desire to modify the implementation itself, we could consider an interface `IShroudedBuffer<T>` to accompany the concrete type.
```